libirods_rule_engine_plugin-audit_amqp.cpp

```diff
@@ -1,413 +1,544 @@
-            ctr_str << ctr;
-            
-            std::string key = elem.first;
-            if (ctr > 1) {
-                key += "__";
-                key += ctr_str.str();
-            }
-
-            json_object_set(
-                obj,
-                key.c_str(),
-                json_string(elem.second.c_str()));
-
-            ++ctr; 
-            ctr_str.clear();
-
-        } // for elem
-    } // for itr
-
-    char* tmp_buf = json_dumps( obj, JSON_INDENT( 0 ) );
-    std::string msg_str = std::string("__BEGIN_JSON__") + std::string(tmp_buf) + std::string("__END_JSON__");
-
-    pn_message_t * message;
-    pn_data_t * body;
-
-    message = pn_message();
-
-    std::string address = audit_amqp_location + "/" + audit_amqp_topic;
-
-    pn_message_set_address(message, address.c_str());
-    body = pn_message_body(message);
-    pn_data_put_string(body, pn_bytes(msg_str.length(), msg_str.c_str()));
-    pn_messenger_put(messenger, message);
-    pn_messenger_send(messenger, -1);
-    
-    pn_message_free(message);
-
-    free(tmp_buf);
-    json_decref(obj);
-
-    if (test_mode) {
-        log_file_ofstream << msg_str << std::endl;
-    }
-
-    return err;
+auto exec_rule(
+	[[maybe_unused]] irods::default_re_ctx& re_ctx,
+	const std::string& _rn,
+	std::list<boost::any>& _ps,
+	irods::callback _eff_hdlr) -> irods::error
+{
+	std::lock_guard<std::mutex> lock(audit_plugin_mutex);
+
+	// stores a counter of unique arg types
+	std::map<std::string, int> arg_type_map;
+
+	ruleExecInfo_t* rei = nullptr;
+	irods::error err = _eff_hdlr("unsafe_ms_ctx", &rei);
+	if (!err.ok()) {
+		return err;
+	}
+
+	nlohmann::json json_obj;
+
+	std::string msg_str;
+	std::string log_file;
+
+	try {
+		uint64_t time_ms = ts_clock::now().time_since_epoch() / std::chrono::milliseconds(1);
+		json_obj["@timestamp"] = time_ms;
+
+		std::string host_name = boost::asio::ip::host_name();
+		json_obj["hostname"] = host_name;
+
+		json_obj["pid"] = getpid();
+
+		json_obj["rule_name"] = _rn;
+
+		for (const auto& itr : _ps) {
+			// The BytesBuf parameter should not be serialized because this commonly contains
+			// the entirety of the contents of files. These could be very big and cause the
+			// message broker to explode.
+			if (std::type_index(typeid(BytesBuf*)) == std::type_index(itr.type())) {
+				// clang-format off
+				log_re::trace({
+					{"rule_engine_plugin", rule_engine_name},
+					{"log_message", "skipping serialization of BytesBuf parameter"},
+					{"rule_name", _rn},
+				});
+				// clang-format on
+				continue;
+			}
+
+			// serialize the parameter to a map
+			irods::re_serialization::serialized_parameter_t param;
+			irods::error ret = irods::re_serialization::serialize_parameter(itr, param);
+			if (!ret.ok()) {
+				// clang-format off
+				log_re::error({
+					{"rule_engine_plugin", rule_engine_name},
+					{"log_message", "unsupported argument for calling re rules from the rule language"},
```

Comment:
`param` does not contain the name of the parameter, as far as I can tell. Even if it did, we wouldn't be able to trust it here since `serialize_parameter` failed.